pkg/relay/ethereum/chain.go

```diff
@@ -269,11 +269,29 @@ func (c *Chain) QueryUnreceivedPackets(ctx core.QueryContext, seqs []uint64) ([]
 	logger := c.GetChannelLogger()
 	var ret []uint64
 	for _, seq := range seqs {
-		found, err := c.ibcHandler.HasPacketReceipt(c.callOptsFromQueryContext(ctx), c.pathEnd.PortID, c.pathEnd.ChannelID, seq)
-		if err != nil {
-			logger.Error("failed to get packet receipt", err)
-			return nil, err
-		} else if !found {
+		var received bool
+
+		// With UNORDERED channel, we can use receipts to check if packets are already received.
+		// With ORDERED channel, since IBC impls don't record receipts, we need to check nextSequenceRecv.
+		switch c.Path().GetOrder() {
+		case chantypes.UNORDERED:
+			var err error
+			if received, err = c.ibcHandler.HasPacketReceipt(c.callOptsFromQueryContext(ctx), c.pathEnd.PortID, c.pathEnd.ChannelID, seq); err != nil {
+				logger.Error("failed to get packet receipt", err)
+				return nil, err
+			}
+		case chantypes.ORDERED:
+			next, err := c.ibcHandler.GetNextSequenceRecv(c.callOptsFromQueryContext(ctx), c.pathEnd.PortID, c.pathEnd.ChannelID)
```

Comment:
Can we call `GetNextSequenceRecv` once outside for-loop?

Reply:
Thank you, it's smart. I will fix it immediately.